File: src/main/java/cz/solutions/cockroach/CockroachMain.java
package cz.solutions.cockroach;

import com.google.common.base.Charsets;
import com.google.common.io.CharStreams;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.text.MessageFormat;
import java.util.logging.Logger;


public class CockroachMain {
    private static final Logger LOGGER = Logger.getLogger(CockroachMain.class.getName());
    public void report(File schwabExportFile, int year, File outputDir) throws IOException {

        ReportGenerator reportGenerator = new ReportGenerator();

        ParsedExport parsedExport = parseExportFile(schwabExportFile);
        Report fixedRateReport = reportGenerator.generateForYear(parsedExport, year, YearConstantExchangeRateProvider.hardcoded());
        Report dynamicRateReport = reportGenerator.generateForYear(parsedExport, year, TabularExchangeRateProvider.hardcoded());


        reportOneVariant(year, outputDir, fixedRateReport,"fixed");
        reportOneVariant(year, outputDir, dynamicRateReport,"dynamic");

        recommendBetterAlternative(fixedRateReport,dynamicRateReport,true);
        recommendBetterAlternative(fixedRateReport,dynamicRateReport,false);



        System.out.println("######################################################");
        System.out.println("# Let's see the guide in" + new File(outputDir, "<fixed/dynamic>_guide_" + year + ".html").getAbsolutePath());
        System.out.println("######################################################");

    }

    private void reportOneVariant(int year, File outputDir,Report data,String dollarConversionSchema) throws IOException {
        FileUtils.writeStringToFile(new File(outputDir, dollarConversionSchema+"_dividend_" + year + ".md"), data.getDividend(), StandardCharsets.UTF_8);
        FileUtils.writeStringToFile(new File(outputDir, dollarConversionSchema+"_rsu_" + year + ".md"), data.getRsu(), StandardCharsets.UTF_8);
        FileUtils.writeStringToFile(new File(outputDir, dollarConversionSchema+"_espp_" + year + ".md"), data.getEspp(), StandardCharsets.UTF_8);
        FileUtils.writeStringToFile(new File(outputDir, dollarConversionSchema+"_sales_" + year + ".md"), data.getSales(), StandardCharsets.UTF_8);
        FileUtils.writeStringToFile(new File(outputDir, dollarConversionSchema+"_guide_" + year + ".html"), data.getGuide(), StandardCharsets.UTF_8);
    }

    private ParsedExport parseExportFile(File schwabExportFile) throws IOException {
        String extension = FilenameUtils.getExtension(schwabExportFile.getName());

        if(extension.equals("json")) {
            JsonExportParser exportParser = new JsonExportParser();
            return exportParser.parse(load(schwabExportFile));
        }else{
            throw new IllegalArgumentException("only .json  files are supported");
        }

    }

    private void recommendBetterAlternative(Report fixedRateReport ,Report dynamicRateReport, boolean use2024Legislative) {
        double profitWhenUsedFixedRate = fixedRateReport.rsuAndEsppAndSalesProfitCroneValue(use2024Legislative);
        double profitWhenUsedDynamicRate = dynamicRateReport.rsuAndEsppAndSalesProfitCroneValue(use2024Legislative);
        String recommendation;
        if(profitWhenUsedFixedRate<=profitWhenUsedDynamicRate){
            recommendation= MessageFormat.format(
                    "Use fixed Dollar conversion rate, because {0}<={1} (diff={2})",
                    FormatingHelper.formatDouble(profitWhenUsedFixedRate),
                    FormatingHelper.formatDouble(profitWhenUsedDynamicRate),
                    FormatingHelper.formatDouble(profitWhenUsedDynamicRate-profitWhenUsedFixedRate)
            );

        } else{
            recommendation= MessageFormat.format(
                    "Use dynamic Dollar conversion rate, because {0}<{1} (diff={2})",
                    FormatingHelper.formatDouble(profitWhenUsedDynamicRate),
                    FormatingHelper.formatDouble(profitWhenUsedFixedRate),
                    FormatingHelper.formatDouble(profitWhenUsedFixedRate-profitWhenUsedDynamicRate)
            );
        }

        System.out.println("######################################################");
        if(use2024Legislative) {
            System.out.println("# Recommendation with 2024 legislative: ");
        }else{
            System.out.println("# Recommendation with old legislative: ");
        }
        System.out.println("# " + recommendation);
        System.out.println("######################################################");
        System.out.printf("\n");
    }
    public static String load(File file) {
        try {
            InputStream is = FileUtils.openInputStream(file);
            return CharStreams.toString(new InputStreamReader(is, Charsets.UTF_8));
        } catch (Exception e) {
            throw new RuntimeException("Could not load file " + file.getAbsolutePath(), e);
        }
    }
    public static void main(String[] args) throws IOException {
        CockroachMain cockroachMain = new CockroachMain();
        cockroachMain.report(new File(args[0]), Integer.parseInt(args[1]), new File(args[2]));
    }
}

File: src/main/kotlin/cz/solutions/cockroach/CockroachMain.kt
package cz.solutions.cockroach

import java.io.File
import java.nio.charset.StandardCharsets
import kotlin.system.exitProcess

fun main(args: Array<String>) {
    try {
        runCockroach(args)
    } catch (e: IllegalArgumentException) {
        System.err.println("Error: ${e.message}")
        exitProcess(1)
    }
}

private fun runCockroach(args: Array<String>) {
    if (args.size == 1 && (args[0].endsWith(".yaml") || args[0].endsWith(".yml"))) {
        val config = CockroachConfig.load(File(args[0]))
        val sources = buildList {
            config.schwab?.let { add(SchwabBrokerSource(File(it))) }
            if (config.etrade != null || config.etradeBenefitHistory != null) {
                add(ETradeBrokerSource(
                    directory = config.etrade?.let { File(it) },
                    benefitHistoryFile = config.etradeBenefitHistory?.let { File(it) },
                ))
            }
            if (config.degiro.isNotEmpty()) add(DegiroBrokerSource(config.degiro.map { File(it) }))
            if (config.revolut.stocks.isNotEmpty() || config.revolut.savings.isNotEmpty()) {
                add(RevolutBrokerSource(
                    stocksFiles = config.revolut.stocks.map { File(it) },
                    savingsFiles = config.revolut.savings.map { File(it) },
                    whtRate = config.revolut.whtRate,
                ))
            }
            if (config.etoro.isNotEmpty()) add(EtoroBrokerSource(config.etoro.map { File(it) }))
            if (config.vub.isNotEmpty()) add(VubBrokerSource(config.vub.map { File(it) }, year = config.year))
        }
        CockroachMain.report(config.year, File(config.outputDir), sources)
        return
    }
    if (args.size < 3) {
        printUsage()
        exitProcess(1)
    }
    val sources = buildList {
        add(SchwabBrokerSource(File(args[0])))
        if (args.size > 3) add(ETradeBrokerSource(directory = File(args[3])))
    }
    CockroachMain.report(args[1].toInt(), File(args[2]), sources)
}

private fun printUsage() {
    System.err.println("Usage: cockroach <config.yaml>                                              (recommended)")
    System.err.println("       cockroach <schwab-json-export> <year> <output-dir> [etrade-dir]     (Schwab/E-Trade only)")
    System.err.println()
    System.err.println("Positional CLI form (limited to E-Trade and Schwab):")
    System.err.println("  schwab-json-export  Path to the Schwab JSON export file")
    System.err.println("  year                Tax year (e.g. 2025)")
    System.err.println("  output-dir          Directory for generated reports")
    System.err.println("  etrade-dir          Optional E-Trade data directory with subdirs:")
    System.err.println("                        rsu/        - RSU release confirmation PDFs")
    System.err.println("                        espp/       - ESPP purchase confirmation PDFs")
    System.err.println("                        dividends/  - single dividends XLSX file")
    System.err.println("                        sales/      - single Gain & Loss CSV/XLSX file")
    System.err.println()
    System.err.println("YAML config form (supports every broker):")
    System.err.println("  year:                  Tax year, e.g. 2025")
    System.err.println("  outputDir:             Directory for generated reports")
    System.err.println("  schwab:                Path to a Schwab JSON export (optional)")
    System.err.println("  etrade:                Path to an E-Trade data directory (optional)")
    System.err.println("  etradeBenefitHistory:  Path to an E-Trade benefit-history XLSX (optional)")
    System.err.println("  degiro:                List of Degiro account-statement CSV paths (optional)")
    System.err.println("  revolut.stocks:        List of Revolut stock statement paths (optional)")
    System.err.println("  revolut.savings:       List of Revolut savings statement paths (optional)")
    System.err.println("  revolut.whtRate:       Withholding-tax rate applied to Revolut dividends (optional)")
    System.err.println("  etoro:                 List of eToro XLSX export paths (optional)")
    System.err.println("  vub:                   List of VÚB interest-confirmation PDF paths (optional)")
    System.err.println()
    System.err.println("At least one broker source must be configured.")
}

object CockroachMain {

    fun report(year: Int, outputDir: File, sources: List<BrokerSource>) {
        require(sources.isNotEmpty()) {
            "No input sources provided. Specify at least one of: schwab, etrade, degiro, revolut, etoro, vub."
        }
        val parsedExport = sources
            .map { it.parse() }
            .fold(ParsedExport.empty()) { acc, e -> acc + e }
        val dailyRateProvider = TabularExchangeRateProvider.fromSource(
            ClasspathOrHttpCnbYearRatesSource(HttpCnbYearRatesSource(HttpCnbYearRatesSource.defaultCacheDir())),
            (year - 1)..year
        )
        val fixedRateReport = ReportGenerator.generateForYear(parsedExport, year, YearConstantExchangeRateProvider.hardcoded())
        val dynamicRateReport = ReportGenerator.generateForYear(parsedExport, year, dailyRateProvider)

        reportOneVariant(year, outputDir, fixedRateReport, "fixed")
        reportOneVariant(year, outputDir, dynamicRateReport, "dynamic")

        recommendBetterAlternative(fixedRateReport, dynamicRateReport)


        println("######################################################")
        println("# Let's see the guide in ${File(outputDir, "<fixed/dynamic>_guide_$year.html").absolutePath}")
        println("######################################################")
    }

    private fun reportOneVariant(year: Int, outputDir: File, data: Report, dollarConversionSchema: String) {
        File(outputDir, "${dollarConversionSchema}_dividend_$year.pdf").writeBytes(data.getDividendPdf())
        File(outputDir, "${dollarConversionSchema}_interest_$year.pdf").writeBytes(data.getInterestPdf())
        File(outputDir, "${dollarConversionSchema}_rsu_$year.pdf").writeBytes(data.getRsuPdf())
        File(outputDir, "${dollarConversionSchema}_espp_$year.pdf").writeBytes(data.getEsppPdf())
        File(outputDir, "${dollarConversionSchema}_sales_$year.pdf").writeBytes(data.getSalesPdf())
        File(outputDir, "${dollarConversionSchema}_guide_$year.html").writeText(data.getGuide(), StandardCharsets.UTF_8)

        val transitionYear = ReportGenerator.LEGISLATIVE_TRANSITION_YEAR
        File(outputDir, "${dollarConversionSchema}_rsu_$transitionYear.pdf").writeBytes(data.getRsu2024Pdf())
        File(outputDir, "${dollarConversionSchema}_espp_$transitionYear.pdf").writeBytes(data.getEspp2024Pdf())
    }



    private fun recommendBetterAlternative(fixedRateReport: Report, dynamicRateReport: Report) {
        val profitWhenUsedFixedRate = fixedRateReport.rsuAndEsppAndSalesProfitCroneValue()
        val profitWhenUsedDynamicRate = dynamicRateReport.rsuAndEsppAndSalesProfitCroneValue()

        val profit2024WhenUsedFixedRate = fixedRateReport.rsuAndEsppProfitCroneValue2024()
        val profit2024WhenUsedDynamicRate = dynamicRateReport.rsuAndEsppProfitCroneValue2024()

        val recommendationOldLegislativeUsedIn2024 = if (profitWhenUsedFixedRate <= profitWhenUsedDynamicRate) {
            "Use fixed Dollar conversion rate, because ${FormatingHelper.formatDouble(profitWhenUsedFixedRate)}<=${FormatingHelper.formatDouble(profitWhenUsedDynamicRate)} (diff=${FormatingHelper.formatDouble(profitWhenUsedDynamicRate - profitWhenUsedFixedRate)})"
        } else {
            "Use dynamic Dollar conversion rate, because ${FormatingHelper.formatDouble(profitWhenUsedDynamicRate)}<${FormatingHelper.formatDouble(profitWhenUsedFixedRate)} (diff=${FormatingHelper.formatDouble(profitWhenUsedFixedRate - profitWhenUsedDynamicRate)})"
        }

        val transitionYear = ReportGenerator.LEGISLATIVE_TRANSITION_YEAR
        println("######################################################")
        println("# Recommendation (If old legislative was used for $transitionYear): ")
        println("# $recommendationOldLegislativeUsedIn2024")
        println("######################################################")
        println()



        val recomendationNewLegislativeUsed2024 = if (profit2024WhenUsedFixedRate+profitWhenUsedFixedRate <= profit2024WhenUsedDynamicRate+profitWhenUsedDynamicRate) {
            "Use fixed Dollar conversion rate  because " +
                    "${FormatingHelper.formatDouble(profit2024WhenUsedFixedRate)}+${FormatingHelper.formatDouble(profitWhenUsedFixedRate)}<=" +
                    "${FormatingHelper.formatDouble(profit2024WhenUsedDynamicRate)}+${FormatingHelper.formatDouble(profitWhenUsedDynamicRate)} " +
                    "(diff=${FormatingHelper.formatDouble(profit2024WhenUsedDynamicRate+profitWhenUsedDynamicRate - profit2024WhenUsedFixedRate-profitWhenUsedFixedRate)})"
        } else {
            "Use dynamic Dollar conversion rate, because " +
                    "${FormatingHelper.formatDouble(profit2024WhenUsedDynamicRate)}+${FormatingHelper.formatDouble(profitWhenUsedDynamicRate)}" +
                    "<${FormatingHelper.formatDouble(profit2024WhenUsedFixedRate)} +${FormatingHelper.formatDouble(profitWhenUsedFixedRate)} " +
                    "(diff=${FormatingHelper.formatDouble(profit2024WhenUsedFixedRate +profitWhenUsedFixedRate  - profit2024WhenUsedDynamicRate-profitWhenUsedDynamicRate)})"
        }

        println("######################################################")
        println("# Recommendation (If new legislative was used in $transitionYear) ")
        println("# $recomendationNewLegislativeUsed2024")
        println("######################################################")

    }
}